adapters/adhese/adhese.go

```diff
@@ -0,0 +1,274 @@
+	} else {
+		var openRtbBidResponseArray []openrtb_ext.AdheseOpenRtbBid
+		if err := json.Unmarshal(response.Body, &openRtbBidResponseArray); err != nil {
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed as Adhese OpenRtb Bid.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseOpenRtbBid(openRtbBidResponseArray[0])
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{WrapError("Response resulted in an empty seatBid array.")}
+	}
+
+	var errs []error
+	for _, sb := range bidResponse.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getBidType(bid.AdM),
+			})
+
+		}
+	}
+	return bidderResponse, errs
+}
+
+func convertAdheseBid(adheseBid openrtb_ext.AdheseBid) openrtb.BidResponse {
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	adheseObj, _ := json.Marshal(openrtb_ext.ExtAdhese{
+		CreativeId:                adheseBid.Id,
+		AdFormat:                  adheseBid.AdFormat,
+		AdType:                    adheseBid.AdType,
+		AdspaceId:                 adheseBid.AdspaceId,
+		DealId:                    adheseBid.OrderId,
+		LibId:                     adheseBid.LibId,
+		OrderProperty:             adheseBid.OrderProperty,
+		Priority:                  adheseBid.Priority,
+		ViewableImpressionCounter: adheseBid.ViewableImpressionCounter,
+	})
+	return openrtb.BidResponse{
+		ID: adheseBid.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				ID:     adheseBid.Id,
+				ImpID:  adheseBid.CreativeName,
+				DealID: adheseBid.OrderId,
+				Price:  price,
+				W:      width,
+				H:      height,
+				CID:    adheseBid.OrderId,
+				CrID:   adheseBid.Id,
+				BURL:   adheseBid.Tracker,
+				AdM:    getAdMarkup(adheseBid),
+				Ext:    adheseObj,
+			}},
+			Seat: "",
+		}},
+		BidID: adheseBid.OrderId,
+		Cur:   adheseBid.Extension.Prebid.Cpm.Currency,
+	}
+}
+
+func convertAdheseOpenRtbBid(adheseBid openrtb_ext.AdheseOpenRtbBid) openrtb.BidResponse {
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	var response openrtb.BidResponse = adheseBid.OriginData
+	response.ID = adheseBid.Origin
+	if adheseBid.OriginInstance != "" {
+		response.ID = response.ID + "-" + adheseBid.OriginInstance
+	}
+	if len(response.SeatBid) > 0 && len(response.SeatBid[0].Bid) > 0 {
+		response.SeatBid[0].Bid[0].Price = price
+		response.SeatBid[0].Bid[0].W = width
+		response.SeatBid[0].Bid[0].H = height
+		response.SeatBid[0].Bid[0].AdM = adheseBid.Body
+		if ContainsAny(adheseBid.Body, []string{"<script", "<div", "<html"}) {
```

Comment:
`adheseBid.ImpressionCounter` shouldn't ever be there for open market responses. Gateway adds the pixel itself. the whole if should go.

Reply:
Also can a similar syntax as above be used to create these nested fields?
By the syntax above, I mean:
```
		SeatBid: []openrtb.SeatBid{{
			Bid: []openrtb.Bid{{
				ID:     adheseBid.Id,
				ImpID:  adheseBid.CreativeName,
				DealID: adheseBid.OrderId,
```
That looks much better.

Reply:
It does doesn't it. But one is a constructor  -I'll try it :)

Reply:
Looks like it won't allow that shorthand for non-types